cli/data.go

```diff
@@ -0,0 +1,183 @@
+package cli
+
+import (
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/pkg/errors"
+	datapb "go.viam.com/api/app/data/v1"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+const (
+	dataDir     = "data"
+	metadataDir = "metadata"
+)
+
+// BinaryData downloads binary data matching filter to dst.
+func (c *AppClient) BinaryData(dst string, filter *datapb.Filter) error {
+	if err := c.ensureLoggedIn(); err != nil {
+		return err
+	}
+
+	if err := makeDestinationDirs(dst); err != nil {
+		return errors.Wrapf(err, "error creating destination directories")
+	}
+
+	skip := int64(0)
+	for {
+		resp, err := c.dataClient.BinaryDataByFilter(context.Background(), &datapb.BinaryDataByFilterRequest{
+			DataRequest: &datapb.DataRequest{
+				Filter: filter,
+				Skip:   skip,
+				Limit:  1,
+			},
+			IncludeBinary: true,
+			CountOnly:     false,
+		})
+		if err != nil {
+			return errors.Wrapf(err, "received error from server")
+		}
+		data := resp.GetData()
+
+		// If no data is returned, there is no more data.
+		if len(data) == 0 {
+			break
+		}
+
+		if len(data) != 1 {
+			return errors.Errorf("expected a single response, received %d", len(data))
+		}
+
+		datum := data[0]
+		mdJSONBytes, err := protojson.Marshal(datum.GetMetadata())
+		if err != nil {
+			return err
+		}
+
+		//nolint:gosec
+		jsonFile, err := os.Create(filepath.Join(dst, "metadata", datum.GetMetadata().GetId()+".json"))
+		if err != nil {
+			return err
+		}
+		if _, err := jsonFile.Write(mdJSONBytes); err != nil {
+			return err
+		}
+
+		gzippedBytes := datum.GetBinary()
+		r, err := gzip.NewReader(bytes.NewBuffer(gzippedBytes))
+		if err != nil {
+			return err
+		}
+
+		//nolint:gosec
+		dataFile, err := os.Create(filepath.Join(dst, "data", datum.GetMetadata().GetId()+datum.GetMetadata().GetFileExt()))
+		if err != nil {
+			return errors.Wrapf(err, fmt.Sprintf("error creating file for file %s", datum.GetMetadata().GetId()))
+		}
+		//nolint:gosec
+		if _, err := io.Copy(dataFile, r); err != nil {
+			return err
+		}
+		if err := r.Close(); err != nil {
+			return err
+		}
+		skip++
+	}
+
+	return nil
+}
+
+// TabularData downloads binary data matching filter to dst.
+func (c *AppClient) TabularData(dst string, filter *datapb.Filter) error {
```

Comment:
note: This might need to be tweaked in a follow-up PR when https://github.com/viamrobotics/app/pull/793 is merged, depending on results of actually using it